### ISSUE TITLE
python image is too large

### ISSUE THREAD

**Issue by ahmelsayed** [Azure/azure-functions-docker#2]

**Comment by davidebbo** [comment]
That's a bit vague :)

**Comment by asavaritayal** [comment]
The image is ~800MB. The (preview sample)[https://github.com/Azure/azure-functions-docker-python-sample] is 1GB. Definitely needs attention :(

**Comment by ahmelsayed** [comment]
It's 400ish now https://hub.docker.com/r/microsoft/azure-functions-python3.6/tags/ 

The base image `azure-functions-base` is 150ish, and the python image which just installs python on top of that image and adds the worker (500 KBs) is 400. That's 250 MBs for python alone which matches the python non-slim image sizes from the official python images https://hub.docker.com/r/library/python/tags/ 

@asavaritayal  can we check with the python devs the possibility of using a slim version of python images? I'm not too sure what the difference is. The python repo docs [highly recommends using the regular image over the slim one](https://github.com/docker-library/docs/tree/master/python#pythonslim) but it doesn't go into too much details other than the slim image missing some packages.

**Comment by asavaritayal** [comment]
> Can we check with the python devs the possibility of using a slim version of python images? I'm not too sure what the difference is. The python repo docs highly recommends using the regular image over the slim one but it doesn't go into too much details other than the slim image missing some packages.

@brettcannon, @1st1, @elprans - to suggest. Currently we're using the `python:3.6.4-jessie` image.

**Comment by brettcannon** [comment]
Since this is Docker's image and not an official one from the Python team I can't directly speak to why they make that recommendation. If I had to guess, though, it's because the slim image might lack stuff like clang to compile C extensions or other common C libraries which some C extensions might expect to be there. Otherwise my Twitter poll suggests most Python developers use Alpine in production: https://snarky.ca/what-are-the-popular-docker-images-for-python/ .

**Comment by elprans** [comment]
Alpine is not compatible with manylinux though, since it's not glibc-based.

**Comment by brettcannon** [comment]
@elprans correct, and I asked some folks who preferred Alpine and they said they didn't care. Now it's going to be a different situation here for installing dependencies versus installing a wheel when bringing your own code, but it seems those who use their own containers have a preference.

**Comment by elprans** [comment]
@brettcannon For simple stuff alpine works perfectly. Complications arise when you need a binary package which requires a ton of `-dev` packages to compile from source, and you end up polluting your image anyway.  And the UX is not great, since it's not always obvious which package you need to install in order to resolve some `gcc: foo.h: no such file or directory`.

Basically, we are trading compatibility and ease of use for image size.  

Is offering multiple image variants an option?

**Comment by brettcannon** [comment]
@elprans it probably isn't worth the headache. I suspect it would be better to look at what is missing in a slim versus full image or look at alternatives like Debian or Fedora images.

**Comment by elprans** [comment]
@brettcannon `-slim` doesn't include any `-dev` packages, so it's equivalent to `-alpine` in many ways.

**Comment by brettcannon** [comment]
Then this is getting down to comparing the image size between various Linux distros to see who has a smaller server image (unless we can convince PyPI or Microsoft to host a wheel-building farm for every package uploaded to PyPI 😉 ).
